File: backend/app/schemas/template.py
"""
Pydantic schemas for the Template resource.

Three shapes:
    TemplateCreate    - what the client sends to POST /templates
    TemplateUpdate    - what the client sends to PATCH /templates/{id}
    TemplateResponse  - what the server returns for GET /templates/{id}
"""

from datetime import datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.common import (
    ORMBase,
    IdentifierStr,
    DisplayNameStr,
    DescriptionStr,
)


# --------------------------------------------------
# Create
# --------------------------------------------------
# Sent to POST /templates to create a new Draft template.
# Only the fields the user provides in the first wizard step.
# The rest (id, status, created_by, etc.) are server generated.

class TemplateCreate(BaseModel):
    """Request body for creating a new template."""

    name: IdentifierStr = Field(
        ...,
        description="Technical name - used as table name. Lowercase alphanumeric + underscore, must start with a letter."
    )
    display_name: DisplayNameStr = Field(
        ...,
        description="Human-readable name shown in the UI."
    )
    description: Optional[DescriptionStr] = Field(
        None,
        description="Optional description of what this template is for."
    )
    domain_id: UUID = Field(
        ...,
        description="ID of the domain this template belongs to."
    )


# --------------------------------------------------
# Update
# --------------------------------------------------
# Sent to PATCH /templates/{id}. All fields optional - client
# only sends the ones being updated. A field not present in
# the request is left unchanged in the database.

class TemplateUpdate(BaseModel):
    """Request body for updating a template."""

    display_name: Optional[DisplayNameStr] = None
    description: Optional[DescriptionStr] = None
    file_format: Optional[str] = Field(
        None,
        pattern=r"^(csv|xlsx)$",
        description="File format accepted for uploads."
    )
    delimiter: Optional[str] = Field(
        None,
        max_length=1,
        description="Delimiter character for CSV uploads."
    )
    encoding: Optional[str] = Field(
        None,
        max_length=20,
        description="Character encoding."
    )
    write_mode: Optional[str] = Field(
        None,
        pattern=r"^(append|overwrite)$",
        description="How uploads are written to the Delta table."
    )
    bad_row_threshold: Optional[Decimal] = Field(
        None,
        ge=0,
        le=100,
        description="Maximum percentage of bad rows allowed (0-100)."
    )
    bad_row_action: Optional[str] = Field(
        None,
        pattern=r"^(drop|fail)$",
        description="What to do with bad rows under the threshold."
    )
    reader_group: Optional[str] = Field(
        None,
        max_length=255,
        description="Databricks group to grant SELECT access to."
    )


# --------------------------------------------------
# Response
# --------------------------------------------------
# Returned from GET /templates/{id} and GET /templates.
# Includes server-generated fields like id, status, timestamps.

class TemplateResponse(ORMBase):
    """Response shape for Template endpoints."""

    id: UUID
    name: str
    display_name: str
    description: str | None
    domain_id: UUID
    uc_table_name: str
    fully_qualified_name: str
    file_format: str
    delimiter: str | None
    encoding: str
    write_mode: str
    bad_row_threshold: Decimal
    bad_row_action: str
    reader_group: str | None
    status: str
    version: int
    parent_template_id: UUID | None
    created_by: str
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/template.ts
/**
 * TypeScript types for the Template resource.
 *
 * Mirrors backend/app/schemas/template.py.
 *
 * A template defines the structure of a manual mapping file
 * and the corresponding Unity Catalog Delta table.
 */

/**
 * Lifecycle status values.
 *
 * Must stay in sync with both:
 *   - SQL CHECK constraint in sql/03_create_templates_table.sql
 *   - app/models/template.py default value
 */
export type TemplateStatus =
  | 'Draft'
  | 'Pending Approval'
  | 'Pending DDL'
  | 'Approved'
  | 'Rejected'
  | 'DDL Failed'
  | 'Deprecated'

export type FileFormat = 'csv' | 'xlsx'
export type WriteMode = 'append' | 'overwrite'
export type BadRowAction = 'fail' | 'drop'

/** Full response shape returned by GET /templates/{id}. */
export interface Template {
  id: string
  name: string
  display_name: string
  description: string | null

  domain_id: string

  uc_table_name: string
  fully_qualified_name: string

  file_format: FileFormat
  delimiter: string
  encoding: string
  write_mode: WriteMode

  /** Percentage 0-100. Returned as a string by Pydantic Decimal. */
  bad_row_threshold: string
  bad_row_action: BadRowAction

  storage_path: string | null
  reader_group: string | null

  status: TemplateStatus
  version: number
  parent_template_id: string | null

  created_by: string
  created_at: string
  updated_at: string
  approved_at: string | null
  databricks_ddl_run_id: string | null
}

/**
 * Payload for POST /templates.
 *
 * The user provides only the basic identification fields; the
 * server fills in defaults for everything else.
 */
export interface TemplateCreate {
  name: string
  display_name: string
  description?: string | null
  domain_id: string
}

/**
 * Payload for PATCH /templates/{id}.
 *
 * All fields are optional — only fields the client actually wants
 * to change need to be sent. Backend uses model_dump(exclude_unset=True)
 * so missing fields are NOT touched.
 */
export interface TemplateUpdate {
  display_name?: string
  description?: string | null
  file_format?: FileFormat
  delimiter?: string
  encoding?: string
  write_mode?: WriteMode
  bad_row_threshold?: string
  bad_row_action?: BadRowAction
  reader_group?: string | null
}